nomad/container.py

```diff
@@ -306,57 +350,36 @@ def _setup_shares(self):
         for k in existing_shares:
             del container.devices[k]
 
-        # LXD uses user namespaces when running safe containers. This means that it maps a set of
-        # uids and gids on the host to a set of uids and gids in the container.
-        # When considering unprivileged containers we want to ensure that "root user" of such
-        # containers have the proper rights to write in shared folders. To do so we have to retrieve
-        # the UserID on the host-side that is mapped to the "root"'s UserID on the guest-side. This
-        # will allow to set ACL on the host-side for this UID. By doing this we will also allow
-        # "root" user on the guest-side to read/write in shared folders.
-        host_root_uid = None
-        if not self.is_privileged:
-            container_path = os.path.join(get_lxd_dir(), 'containers', self.lxd_name, 'rootfs')
-            container_path_stats = os.stat(container_path)
-            host_root_uid = container_path_stats.st_uid
-
         for i, share in enumerate(self.options.get('shares', []), start=1):
             source = os.path.join(self.homedir, share['source'])
             if source not in existing_sources:
                 logger.info('Setting host-side ACL for {}'.format(source))
-                subprocess.Popen(
-                    'setfacl -Rdm u:{}:rwX {}'.format(os.getuid(), source), shell=True).wait()
-                if host_root_uid is not None:
+                self._host.give_current_user_access_to_share(source)
+                if not self.is_privileged:
                     # We are considering a safe container. So give the mapped root user permissions
                     # to read/write contents in the shared folders too.
-                    subprocess.Popen(
-                        'setfacl -Rm user:lxd:rwx,default:user:lxd:rwx,'
-                        'user:{0}:rwx,default:user:{0}:rwx {1}'.format(host_root_uid, source),
-                        shell=True).wait()
-
-            shareconf = {
-                'type': 'disk',
-                'source': source,
-                'path': share['dest'],
-            }
+                    self._host.give_mapped_user_access_to_share(source)
+                    # We also give these permissions to any user that was created with LXD-Nomad.
+                    for uconfig in self.options.get('users', []):
+                        username = uconfig.get('name')
+                        self._host.give_mapped_user_access_to_share(
```

Comment:
Yes, it works! In an unprivileged container and with some `users:` defined, I could properly access shares both ways. good work!